File: packages/decoder/src/addition_layer.ts
interface WasmExports extends WebAssembly.Exports {
  readonly memory: WebAssembly.Memory;

  init(embedding_size: number): number;
  getInputVector(state: number): number;
  getOutputVector(state: number): number;
  forward(state: number): void;
}

export interface AdditionLayerInit {
  readonly embeddingSize: number;
}

export class AdditionLayer {
  static wasmModule: WebAssembly.Module | undefined;

  static async instantiate(init: AdditionLayerInit): Promise<AdditionLayer> {
    return new AdditionLayer(init, await WebAssembly.instantiate(AdditionLayer.wasmModule!));
  }

  readonly inputVector: Float32Array;
  readonly outputVector: Float32Array;

  readonly #wasmInstance: WebAssembly.Instance;
  readonly #wasmState: number;

  private constructor({ embeddingSize }: AdditionLayerInit, wasmInstance: WebAssembly.Instance) {
    const wasmExports = wasmInstance.exports as WasmExports;
    const wasmState = wasmExports.init(embeddingSize);

    this.inputVector = new Float32Array(
      wasmExports.memory.buffer,
      wasmExports.getInputVector(wasmState),
      embeddingSize,
    );

    this.outputVector = new Float32Array(
      wasmExports.memory.buffer,
      wasmExports.getOutputVector(wasmState),
      embeddingSize,
    );

    this.#wasmInstance = wasmInstance;
    this.#wasmState = wasmState;
  }

  forward(): void {
    const wasmExports = this.#wasmInstance.exports as WasmExports;

    wasmExports.forward(this.#wasmState);
  }
}

File: src/addition_layer.zig
const std = @import("std");

const addition = @import("simd/addition.zig");

const State = struct {
    input_vector: []f32,
    output_vector: []f32,
};

export fn init(embedding_size: usize) ?*const State {
    const allocator = std.heap.page_allocator;
    const state = allocator.create(State) catch return null;

    state.* = .{
        .input_vector = allocator.alloc(f32, embedding_size) catch return null,
        .output_vector = allocator.alloc(f32, embedding_size) catch return null,
    };

    return state;
}

export fn getInputVector(state: *const State) [*]f32 {
    return state.input_vector.ptr;
}

export fn getOutputVector(state: *const State) [*]f32 {
    return state.output_vector.ptr;
}

export fn forward(state: *const State) void {
    addition.compute(state.input_vector, state.output_vector, state.output_vector);
}
